### PR TITLE
LibWeb: clarify empty needle and implement attribute selector sensitivity identifier

### PR DESCRIPTION
**Pull request by therealglazou**
LibWeb: several fixes in Attribute Selector handling:

- `Parser::parse_attribute_simple_selector` was not checking that invalid extra tokens were or not present in the block of brackets representing an attribute selector
- attribute selectors looking for an empty value should result in a false condition on the rule
- Selectors Level 4 case sensitivity identifier (i or s before the closing bracket) were added
- I added a test page of mine, viewable under `More attribute selectors` inside `welcome.html`

### CODE REVIEW

File: Base/res/html/misc/welcome.html [hunk 1]
```diff
@@ -50,6 +50,8 @@ <h1>Welcome to the Serenity Browser!</h1>
     <p>Your user agent is: <b><span id="ua"></span></b></p>
     <p>This page loaded in <b><span id="loadtime"></span></b> ms</p>
     <nav class="page-list">
+        <ul>
+        </ul>
```
Comment (linusg) [inline Base/res/html/misc/welcome.html:54]: What's this for?

File: Base/res/html/misc/welcome.html [hunk 2]
```diff
@@ -86,6 +88,7 @@ <h2>CSS</h2>
             <li><h3>Selectors</h3></li>
             <li><a href="selectors.html">Selectors</a></li>
             <li><a href="attrselectors.html">Attribute selectors</a></li>
+            <li><a href="selectorTest.html#target">More attribute selectors</a></li>
```
Comment (linusg) [inline Base/res/html/misc/welcome.html:91]: Please follow the existing `foo-bar.html` naming scheme

File: Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp [hunk 3]
```diff
@@ -78,23 +78,56 @@ static inline bool matches_checked_pseudo_class(DOM::Element const& element)
 
 static inline bool matches_attribute(CSS::Selector::SimpleSelector::Attribute const& attribute, DOM::Element const& element)
 {
-    switch (attribute.match_type) {
-    case CSS::Selector::SimpleSelector::Attribute::MatchType::HasAttribute:
+    if (attribute.match_type == CSS::Selector::SimpleSelector::Attribute::MatchType::HasAttribute) {
+        // early way out in case of an attribute existence selector
```
Comment (linusg) [inline Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp:82]: Please make sure comments follow the [coding style](https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#comments)

File: Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp [hunk 4]
```diff
@@ -78,23 +78,56 @@ static inline bool matches_checked_pseudo_class(DOM::Element const& element)
...
     case CSS::Selector::SimpleSelector::Attribute::MatchType::ExactValueMatch:
-        return element.attribute(attribute.name) == attribute.value;
-    case CSS::Selector::SimpleSelector::Attribute::MatchType::ContainsWord:
-        return element.attribute(attribute.name).split_view(' ').contains_slow(attribute.value);
+        return case_insensitive_match
+            ? element.attribute(attribute.name).equals_ignoring_case(attribute.value)
+            : element.attribute(attribute.name) == attribute.value;
+    case CSS::Selector::SimpleSelector::Attribute::MatchType::ContainsWord: {
+        if (attribute.value.is_empty()) {
+            // early way out if value is empty
```
Comment (linusg) [inline Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp:98]: Same here

**Review by linusg — CHANGES_REQUESTED** [review]
Please split this up into multiple commits (I think 2 for the code + 1 for the test) - more often than not having 'and' in a commit message indicates it's not atomic enough :)

File: Base/res/html/misc/welcome.html (hunk 1 above)
Comment (therealglazou) [reply Base/res/html/misc/welcome.html:54]: That's a leftover that does not render so I did not see it :( Sorry for that.

File: Userland/Libraries/LibWeb/CSS/Selector.cpp [hunk 5]
```diff
@@ -174,7 +174,18 @@ String Selector::SimpleSelector::serialize() const
 
             serialize_a_string(s, attribute.value);
         }
-        // FIXME: 5. If the attribute selector has the case-sensitivity flag present, append " i" (U+0020 U+0069) to s.
+
+        // 5. serialize the case-sensitivity flag if it is explicitly specified
```
[the remark this answers is not in the document]
Comment (therealglazou) [reply Userland/Libraries/LibWeb/CSS/Selector.cpp:178]: The CSS OM spec is not in sync with Selectors Level 4 wrt to attribute selector case identifiers. It deals only with " i" and not with " s". I'll make a combo from there.

File: Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp (hunk 4 above)
Comment (therealglazou) [reply Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp]: Ok

**Comment by awesomekling** [comment]
Looks great! But you can drop the "LibWeb: In answer to comments by linusg and awesomekling" from the commit message. Looks like an artifact from local squashing :)

File: Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp [hunk 6]
```diff
@@ -78,23 +78,56 @@ static inline bool matches_checked_pseudo_class(DOM::Element const& element)
...
     case CSS::Selector::SimpleSelector::Attribute::MatchType::ExactValueMatch:
-        return element.attribute(attribute.name) == attribute.value;
-    case CSS::Selector::SimpleSelector::Attribute::MatchType::ContainsWord:
-        return element.attribute(attribute.name).split_view(' ').contains_slow(attribute.value);
+        return case_insensitive_match
+            ? element.attribute(attribute.name).equals_ignoring_case(attribute.value)
+            : element.attribute(attribute.name) == attribute.value;
+    case CSS::Selector::SimpleSelector::Attribute::MatchType::ContainsWord: {
+        if (attribute.value.is_empty()) {
+            // Early way out if value is empty.
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/CSS/SelectorEngine.cpp:98]: IMO comments like these don't add anything, it's just saying what the code is doing, which we can gather just by reading the code. Same for the comment on line 82.

**Comment by awesomekling** [comment]
:star:

**Pull request merged**